### ISSUE TITLE
[question] temporal convolutional networks

### ISSUE THREAD

**Issue by antoine-galataud** [hill-a/stable-baselines#138]
Hi,

It's more to have an opinion on the feasibility and relevance of implementing a custom policy network based on temporal convolutional networks (TCN). This [paper](https://bcourses.berkeley.edu/files/70257152/download?download_frd=1) talks about using it with A3C and shows promising results. I couldn't find any other references to such usage in RL though.
It seems like TCNs are gaining traction in supervised learning and seq2seq problems, and there were empirical demonstrations on how they can outperform RNNs in that area (e.g. [here](https://arxiv.org/pdf/1803.01271)). 
I find this interesting also for reducing memory consumption (also compared to RNNs).

I've tested it a bit myself with PPO2. Basically I've replaced LSTM layer of the LstmPolicy with conv1d layers (with specific padding and dilation rate for each layer, there are a few examples of _Temporal blocks_ implementations out there)
Implementation scope wise, I think such custom policy would have the same "limitation" as for LSTM policies: you need to feed it with a sequence of observations, which is not the case (or even possible) with every algo.
For now (but I did only a few runs) my tests didn't show good convergence, but it requires some (lot of) tweaking. PPO2 seems also very sensitive to network architecture.

Would be interesting to hear opinions and/or experiences on that. 
Thx!

**Comment by araffin** [comment]
Hello,

I don't know much about TCN but here are some thoughts on your problem:

>This paper talks about using it with A3C and shows promising results.

I looked at it briefly and it does not seem to follow the best practices for reporting RL results:
- the number of seed used is not clear
- the budget dedicated to hyperparameter tuning for each baseline is not mentioned
- the experiments seems "unfair" for the baselines as they removed the velocity from the observation (breaking the markovian assumption) and did not provide any mechanism like frame-stacking to counter-act this effect for feedforward policies
- they train their algorithms on pong without the common pre-processing step (and again without frame-stacking which is completely unfair for feedforward methods)

As a result, I'd take those results with a grain of salt.

>It seems like TCNs are gaining traction in supervised learning and seq2seq problems, and there were empirical demonstrations on how they can outperform RNNs in that area

I thought that "Attention [was] all you need" now ^^# (cf the transformer and recently BERT from Google)

>For now (but I did only a few runs) my tests didn't show good convergence, but it requires some (lot of) tweaking. PPO2 seems also very sensitive to network architecture.

Well, you changed completely the network architecture, so you need unfortunately to tune a bit the hyperparameters.
Out of curiosity, on which problems are you working on?

**Comment by antoine-galataud** [comment]
Hi,

> I thought that "Attention [was] all you need" now ^^# (cf the transformer and recently BERT from Google)

Well, that's very interesting too. Is integration of attention mechanism on your roadmap yet?

> Out of curiosity, on which problems are you working on?

Industrial applications for which main constraint is energy consumption. Input of model is multivariate time series (from simulator). Action space can be continuous but most of the time discretizing it gives better results. 
The biggest challenge I'm facing so far is that model tends to fit for a single action value after some time (value often optimal though) and doesn't generalize well. So I'm exploring various options to encourage exploration, prevent overfitting, and make the policy "remember" about past experiences.
TRPO was giving the best results so far, but since it doens't make much sense to feed a sequence-based network like TCN with one observation at a time (although it gives interesting results), I'm now testing with PPO2.

**Comment by araffin** [comment]
>Is integration of attention mechanism on your roadmap yet?

It is not for now. We already have a lot of higher priority things I think (see https://github.com/hill-a/stable-baselines/projects/1)

**Comment by antoine-galataud** [comment]
Hi,

There is something I don't get with TRPO: there's a timesteps per batch parameter but it seems there's always only 1 timestep provided for every run (`n_steps` is always = 1). With PPO2, you get n_env * n_steps observations.

Is it fundamentally how TRPO is supposed to work? 

Also it seems that some parameters aren't initialized with TRPO: `n_batch` is `None`.

**Comment by antoine-galataud** [comment]
Nevermind. Just saw it’s being addressed I  #140

**Comment by antoine-galataud** [comment]
Hi,

After some testing, I seem to be able to reproduce what is described in the [paper](https://bcourses.berkeley.edu/files/70257152/download?download_frd=1). Most importantly, sum of rewards converges faster, is higher, and is more stable than with a feed forward network. 

![ff_tcn_sum_of_rewards](https://user-images.githubusercontent.com/5330926/50589402-c4b34400-0e86-11e9-8131-23ae600fcf02.png)

I didn't do a thorough comparison (only tested on my custom env), and my tests aren't fair since same parameters are used for both TCN and FF. 
Also I noticed that results are good with TRPO and PPO1, but not with PPO2 (not even converging), which raises some concerns.
However it feels like a promising option and I'll dedicate more testing time to it.

**Comment by antoine-galataud** [comment]
I'm closing this question as I think:
- I couldn't prove it can be beneficial, and it seems quite unstable
- no state of the art demonstration of how to use it or integrate it (feature extraction layer? somewhere else?)

I would be happy to share some ideas/code portions if anyone interested

**Comment by msinghraniyal** [comment]
@antoine-galataud 

> I'm closing this question as I think:
> 
> * I couldn't prove it can be beneficial, and it seems quite unstable
> * no state of the art demonstration of how to use it or integrate it (feature extraction layer? somewhere else?)
> 
> I would be happy to share some ideas/code portions if anyone interested

I came here looking for TCN implementation, would you be able to share some sample code and your thoughts on it?
